pkg/auth/requests/service_account.go

```diff
@@ -0,0 +1,169 @@
+		return info, hasAuth, nil
+	}
+	clusterID := mux.Vars(req)["clusterID"]
+	if clusterID == "" {
+		return info, hasAuth, fmt.Errorf("no clusterID found in request")
+	}
+	// Check the cluster setting value to determine whether we will continue the auth process
+	settings, err := t.clusterProxyConfigsGetter.List(clusterID, labels.NewSelector())
+	if err != nil {
+		logrus.Debugf("rejecting downstream proxy request for %s, unable to fetch ClusterProxySettings object for cluster", req.URL.Path)
+		return &user.DefaultInfo{}, false, nil
+	}
+	if settings == nil {
+		logrus.Debugf("rejecting downstream proxy request for %s, no ClusterProxySettings object exists for cluster", req.URL.Path)
+		return &user.DefaultInfo{}, false, nil
+	}
+	if len(settings) > 1 {
+		logrus.Errorf("multiple clusterproxyconfigs found for cluster, which is a misconfiguration, feature is disabled")
+		return &user.DefaultInfo{}, false, nil
+	}
+
+	if !settings[0].Enabled {
+		logrus.Debugf("rejecting downstream proxy request for %s, current setting is enabled: %v", req.URL.Path, settings[0].Enabled)
+		return &user.DefaultInfo{}, false, nil
+	}
+
+	// See if the token is a JWT.
+	rawToken := tokens.GetTokenAuthFromRequest(req)
+
+	jwtParser := jwtv4.Parser{}
+	claims := jwtv4.RegisteredClaims{}
+	// Using ParseUnverified is deliberate here to look at the basic info in the token.
+	// Later on, we do a real TokenReview against the downstream cluster to actually verify the JWT.
+	_, _, err = jwtParser.ParseUnverified(rawToken, &claims)
+	if err != nil {
+		logrus.Debug("saauth: error parsing JWT")
+		return info, false, err
+	}
+
+	if !strings.HasPrefix(claims.Subject, serviceaccount.ServiceAccountUsernamePrefix) {
+		logrus.Debugf("saauth: JWT sub is not a service account: %v", err)
+		return info, false, nil
+	}
+
+	if isTokenExpired(claims.ExpiresAt) {
+		logrus.Debugf("saauth: Service Account JWT is expired. Expiration time was: %v", claims.ExpiresAt)
+		return info, false, nil
+	}
+
+	// Get a client for the downstream cluster.
+	downstreamAuthClient, err := t.authClientCreator(clusterID)
+	if err != nil {
+		logrus.Errorf("saauth: failed to fetch downstream kubeconfig: %v", err)
+		return info, false, fmt.Errorf("failed to get downstream auth client when validating token for downstream cluster")
+	}
+
+	tokenReview := &v1.TokenReview{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "authentication.k8s.io/v1",
+			Kind:       "TokenReview",
+		},
+		Spec: v1.TokenReviewSpec{
+			Token: rawToken,
+		},
+	}
+
+	// Make the token review request to the downstream cluster.
+	tokenReview, err = downstreamAuthClient.AuthenticationV1().TokenReviews().Create(req.Context(), tokenReview, metav1.CreateOptions{})
+	if err != nil {
+		logrus.Debugf("saauth: error creating a tokenreview request: %v", err)
+		return info, false, nil
+	}
+
+	// Let others know that this request is authenticated using a service account.
+	*req = *req.WithContext(authcontext.SetSAAuthenticated(req.Context()))
```

Comment:
I suspect that the fact that it modifies the request should be called out in the doc-comment for this method?